src/lib/runway_takeoff/RunwayTakeoff.cpp

```diff
@@ -0,0 +1,286 @@
+		}
+
+		break;
+
+	default:
+		break;
+	}
+}
+
+/*
+ * Returns true as long as we're below navigation altitude
+ */
+bool RunwayTakeoff::controlYaw()
+{
+	// keep controlling yaw directly until we start navigation
+	return _state < RunwayTakeoffState::CLIMBOUT;
+}
+
+/*
+ * Returns pitch setpoint to use.
+ *
+ * Limited (parameter) as long as the plane is on runway. Otherwise
+ * use the one from TECS
+ */
+float RunwayTakeoff::getPitch(float tecsPitch)
+{
+	if (_state <= RunwayTakeoffState::CLAMPED_TO_RUNWAY) {
+		return math::radians(_runway_pitch_sp.get());
+	}
+
+	return tecsPitch;
+}
+
+/*
+ * Returns the roll setpoint to use.
+ */
+float RunwayTakeoff::getRoll(float navigatorRoll)
+{
+	// until we have enough ground clearance, set roll to 0
+	if (_state < RunwayTakeoffState::CLIMBOUT) {
+		return 0.0f;
+	}
+
+	// allow some roll during climbout
+	else if (_state < RunwayTakeoffState::FLY) {
+		return math::constrain(navigatorRoll,
+				       math::radians(-_max_takeoff_roll.get()),
+				       math::radians(_max_takeoff_roll.get()));
+	}
+
+	return navigatorRoll;
+}
+
+/*
+ * Returns the yaw setpoint to use.
+ *
+ * In heading hold mode (_heading_mode == 0), it returns initial yaw as long as it's on the
+ * runway. When it has enough ground clearance we start navigation towards WP.
+ */
+float RunwayTakeoff::getYaw(float navigatorYaw)
+{
+	if (_heading_mode.get() == 0 && _state < RunwayTakeoffState::CLIMBOUT) {
+		return _init_yaw;
+
+	} else {
+		return navigatorYaw;
+	}
+}
+
+/*
+ * Returns the throttle setpoint to use.
+ *
+ * Ramps up in the beginning, until it lifts off the runway it is set to
+ * parameter value, then it returns the TECS throttle.
+ */
+float RunwayTakeoff::getThrottle(float tecsThrottle)
+{
+	switch (_state) {
+	case RunwayTakeoffState::THROTTLE_RAMP: {
+			float throttle = hrt_elapsed_time(&_initialized_time) / (float)_throttle_ramp_time *
+					 _takeoff_throttle.get();
+			return throttle < _takeoff_throttle.get() ?
+			       throttle :
+			       _takeoff_throttle.get();
+		}
+
+	case RunwayTakeoffState::CLAMPED_TO_RUNWAY:
+		return _takeoff_throttle.get();
+
+	default:
+		return tecsThrottle;
+	}
+}
+
+bool RunwayTakeoff::resetIntegrators()
+{
+	// reset integrators if we're still on runway
+	return _state < RunwayTakeoffState::TAKEOFF;
+}
+
+/*
+ * Returns the minimum pitch for TECS to use.
+ *
+ * In climbout we either want what was set on the waypoint (sp_min) but at least
+ * the climbtout minimum pitch (parameter).
+ * Otherwise use the minimum that is enforced generally (parameter).
+ */
+float RunwayTakeoff::getMinPitch(float sp_min, float climbout_min, float min)
+{
+	if (_state < RunwayTakeoffState::FLY) {
+		return math::max(sp_min, climbout_min);
+	}
+
+	else {
+		return min;
+	}
+}
+
+/*
+ * Returns the maximum pitch for TECS to use.
+ *
+ * Limited by parameter (if set) until climbout is done.
+ */
+float RunwayTakeoff::getMaxPitch(float max)
+{
+	// use max pitch from parameter if set (> 0.1)
```

Comment:
@AndreasAntener Could you have a quick look at this? I don't see where the 10degrees are.